grammar: rename match->case

zkat opened the pull request:
Fixes: #68

This renaming removes all outstanding grammar-related concerns
and greatly simplifies things. It also has precedent: Elixir and
Erlang are the two most prominent dynamic languages with pattern
matching support, and are both languages whose semantics this
proposal draws from. It seems perfectly appropriate to use their
name for this construct rather than stick to `match`.

This PR supersedes #72 and blocks #77 from being a possibility. I urge anyone who might have an initial reaction against this name to consider the value behind using an actual keyword with this. Because of the location and context in which this keyword is being used, I do not believe this runs counter the the ongoing design goal of distancing pattern matching from `switch`, which I continue to agree is an important principle behind this design.

/cc @tabatkins @ljharb @isiahmeadows

nijikokun commented:
I'm not sure this change has any improvements over the previous grammar, if anything it makes it more confusing...

ljharb commented:
I am very against this change landing; any similarity to switch imo is extremely problematic. I’m not swayed by arguments to preserve wildly unidiomatic brace styles, nor to avoid ASI hazards.

I also think this change will make teaching/learning harder - suddenly “case” means two things, a part of a discouraged construct, and a whole new encouraged construct of its own.

I would highly recommend not merging this until after committee discussion.

arcanis commented:
I can see the point of using an existing keyword, but (and that's even without considering my proposal from the other issue) using `case` here seems really confusing to me, since they have different meanings in otherwise similar features (sorry @ljharb 😄). Is it the only option? What are the alternatives?

zkat commented:
@ljharb I wanna make sure to make sure this proposal is grammatically sound and that there's very little noise in the signal as far as grammar issues go. From what I'm told, there's a lot of hesitation right now to accept large grammar additions, and so I'm trying to make everything as similar to existing grammar constructs as possible and minimize weird grammar special-casing (like new cover grammars, for example). I believe a proposal that is grammatically more straightforward is significantly much more likely to succeed, and swapping these two words seems like a relatively small sacrifice to make towards that. Again, the thing that pushed me over the wall with this specific change is remembering that the two major *dynamic* languages that use pattern matching both use `case` as their keyword for this, and have literally decades of precedent for it. I agree that those languages also do not have the `case` keyword in `switch`, though.

I understand the argument wrt to staying away from `switch`, though -- please keep in mind how often I've tried to drive that point home in this proposal, and how many other syntactic constructs I've avoided precisely for this purpose, so I'm very much with you on the general intention.

Still, I'm trying to put things on the table that have real, mechanical consequences, because abstract things are harder to justify to a large group of people with widely varying opinions on nuanced concerns. Does that make sense?

I very much intend to let some conversation happen around this change, but please keep in mind that this is the most important change of them all: https://github.com/tc39/proposal-pattern-matching/pull/79/files#diff-04c6e90faac2675aa89e2176d2eec7d8R419

And if this PR is going to be rejected, I would much rather that people opposing the change:

A. acknowledge the concerns outlined in #68 around braces -- I am very much not in the business of throwing people with different brace styles under the bus just 'cause the one I like is the winning one.
B. provide a better alternative for addressing those concerns, or at least provide reassurance that a cover grammar or other parsing backflips aren't going to hold this up in committee. I want the functionality, not necessarily the name.

rattrayalex commented:
> provide a better alternative 

Heh, `private` and `protected` are free and unlikely to be used in the intuitive way 🙃. 

Jokes aside, what are the criteria by which a better name could be derived? I assume the identifier must currently be reserved?
